bookwyrm/utils/isni.py

```diff
@@ -0,0 +1,144 @@
+        "maximumRecords": max_records,
+        "startRecord": "1",
+        "recordPacking": "xml",
+        "sortKeys": "RLV,pica,0,,",
+    }
+    result = requests.get("http://isni.oclc.org/sru/", params=query_params)
+    # the OCLC ISNI server asserts the payload is encoded
+    # in latin1, but we know better
+    result.encoding = "utf-8"
+    return result.text
+
+
+def make_name_string(element):
+    """create a string of form 'personal_name surname'"""
+
+    # NOTE: this will often be incorrect, many naming systems
+    # list "surname" before personal name
+    forename = element.find(".//forename")
+    surname = element.find(".//surname")
+    if forename is not None:
+        return "".join([forename.text, " ", surname.text])
+    return surname.text
+
+
+def get_other_identifier(element, code):
+    """Get other identifiers associated with an author from their ISNI record"""
+
+    identifiers = element.findall(".//otherIdentifierOfIdentity")
+    for section_head in identifiers:
+        if (
+            section_head.find(".//type") is not None
+            and section_head.find(".//type").text == code
+            and section_head.find(".//identifier") is not None
+        ):
+            return section_head.find(".//identifier").text
+    return ""
+
+
+def get_external_information_uri(element, match_string):
+    """Get URLs associated with an author from their ISNI record"""
+
+    sources = element.findall(".//externalInformation")
+    for source in sources:
+        uri = source.find(".//URI")
+        if uri is not None and uri.text.find(match_string) is not None:
+            return uri.text
+    return ""
+
+
+def find_authors_by_name(name_string):
+    """Query the ISNI database for possible author matches by name"""
+
+    payload = request_isni_data("pica.na", name_string)
+    # parse xml
+    root = ET.fromstring(payload)
+    # build list of possible authors
+    possible_authors = []
+    for element in root.iter("responseRecord"):
+
+        personal_name = element.find(".//forename/..")
+        bio = element.find(".//nameTitle")
+
+        if not personal_name:
+            continue
+
+        author = {}
+        author["isni"] = element.find(".//isniUnformatted").text
+        author["uri"] = element.find(".//isniURI").text
+        author["name"] = make_name_string(personal_name)
+        if bio is not None:
+            author["bio"] = bio.text
+        possible_authors.append(author)
+
+    return possible_authors
+
+
+def get_author_isni_data(isni):
+    """Find data to populate a new author record from their ISNI"""
+
+    payload = request_isni_data("pica.isn", isni)
+    # parse xml
+    root = ET.fromstring(payload)
+    # there should only be a single responseRecord
+    # but let's use the first one just in case
+    element = root.find(".//responseRecord")
+    personal_name = element.find(".//forename/..")
+    bio = element.find(".//nameTitle")
+    author = {}
```

Comment:
I might be being super dumb here but how do I get django to stop complaining that I haven't given an `id` key for `activitypub.Author`?

```python
author = models.Author(
  name=name,
  isni=isni,
  viafId=viaf,
  aliases=aliases,
  bio=bio,
  wikipediaLink=wikipedia
)

return author
```

Reply:
You're not being dumb, it's a required field that makes sense in activitypub and less so in this context. If you have the URL it queried, use that as the id, otherwise, use the isni I think. It normally stores the url used to resolve the object, like in: https://bookwyrm.social/author/2601.json

Reply:
might make sense to use the isni.org url, then.

Reply:
yeah, the isni.org url makes sense. the `id` field gets saved as `origin_id` in the model, which can be helpful for tracking down how data got where it is, but isn't actively used for anything

Reply:
Ok I think I mostly understand this and I have a working update, except for one major bummer. If I do this:

```python
author = author_object.to_model(
  model=models.Author, 
  overwrite=False
)
```
...then I don't seem to have any control over which existing author Bookwyrm uses to "merge". In one case I want to merge with any existing author if the ID matches, and in another case we will never want to merge and always want to create a new author. Currently what's happening is it always merges on a matching name, regardless of anything else.

I'm not sure how I can do either of those things: there doesn't seem to be an obvious way to create a new author in the DB, from an activitypub dataclass, nor to force a new author from the dataclass. But you mentioned above that if I do _something_ then the id automagically becomes the `origin_id` so I guess I need to know what that something is?

Apologies, I'm not familiar with dataclasses and I'm struggling to understand the activitypub code. I can see the sense in using the code that's already there to keep things clean and DRY, but I'm missing some key aspect to making this work.

Reply:
Two things to fix this, first is, in `bookwyrm/models/author.py` on line 30, remove `deduplication_field=True` from the `name` field and run a migration, that will stop it merged based on author name -- this was a better-than-nothing quick fix for openlibrary issues that I meant to remove ages ago. To create a new author, you could create an `Author` model object and pass it in with the instance field. If that doesn't work I can modify the `to_model` function to specify requiring a new object.

You shouldn't need to do anything to get the origin_id field set, that happens in the model save function: https://github.com/bookwyrm-social/bookwyrm/blob/main/bookwyrm/models/book.py#L65 if the object is being created.

Truly no apologies necessary, this is the dark underbelly of the code that makes sense only inside my own head. I really appreciate you bearing with me on it, I know it's very unintuitive and does a lot of hand-wavey surprise things in non-obvious places.

Reply:
yesss this rules.

Django doesn't seem to want/need a migration after removing name as a dedupe field. Works like a charm as described, it's a good way to do this, and now I get it. Thanks!